File: 405_Submatrix_Sum.cpp
/*
405. Submatrix Sum
Given an integer matrix, find a submatrix where the sum of numbers is zero. Your code should return the coordinate of the left-up and right-down number.

Example
Given matrix

[
  [1 ,5 ,7],
  [3 ,7 ,-8],
  [4 ,-8 ,9],
]
return [(1,1), (2,2)]

Challenge
O(n^3) time.


Author: Yuan Li
Date: 7/30/2018 
Difficulty: Medium
*/

class Solution {
public:
    /*
     * @param matrix: an integer matrix
     * @return: the coordinate of the left-up and right-down number
     */
    vector<vector<int>> submatrixSum(vector<vector<int>> &matrix) {
        // write your code here
        vector<vector<int>> res;
        
        if (matrix.size() == 0 || matrix[0].size() == 0) {
            return res;
        }
        
        // construct prefix-sum matrix
        vector<vector<int>> prefixSum(matrix.size() + 1, vector<int>(matrix[0].size() + 1, 0));
        for (int r = 1; r < prefixSum.size(); ++r) {
            for (int c = 1; c < prefixSum[0].size(); ++c) {
                prefixSum[r][c] = prefixSum[r-1][c] + prefixSum[r][c-1] - prefixSum[r-1][c-1] +
                                    matrix[r-1][c-1];
            }
        }
        
        // search for submatrix
        for (int r0 = 0; r0 < prefixSum.size() - 1; ++r0) {
            for (int r1 = r0 + 1; r1 < prefixSum.size(); ++r1) {
                unordered_map<int, int> sum2col;
                for (int c = 0; c < prefixSum[0].size(); ++c) {
                    int submatSum = prefixSum[r1][c] - prefixSum[r0][c];
                    if (sum2col.count(submatSum)) {
                        res = {{r0, sum2col[submatSum]}, {r1 - 1, c - 1}};
                        return res;
                    } else {
                        sum2col[submatSum] = c;
                    }
                }
            }
        }
        
        return res;
    }
};

File: 405_Submatrix_Sum.py
#!/usr/bin/python3

'''
405. Submatrix Sum

Given an integer matrix, find a submatrix where the sum of numbers is zero. Your code should return the coordinate of the left-up and right-down number.

If there are multiple answers, you can return any of them.

Example
Example 1:

Input:
[
  [1, 5, 7],
  [3, 7, -8],
  [4, -8 ,9]
]
Output: [[1, 1], [2, 2]]

Example 2:

Input:
[
  [0, 1],
  [1, 0]
]
Output: [[0, 0], [0, 0]]

Challenge
O(n3) time.

Author: Yuan Li
Date: 2/7/2019 
Difficulty: Medium
'''

class Solution:
    """
    @param: matrix: an integer matrix
    @return: the coordinate of the left-up and right-down number
    """
    def submatrixSum(self, matrix):
        # write your code here
        if not matrix or not matrix[0]:
            return None
            
        n, m = len(matrix), len(matrix[0])
        for t in range(n):
            pre_sum_row = [0] * m
            
            for b in range(t, n):
                pre_sum_col = 0
                sum_col = {0: -1}
                
                # subarray sum problem
                for c in range(m):
                    pre_sum_row[c] += matrix[b][c]
                    pre_sum_col += pre_sum_row[c]
                    
                    if pre_sum_col in sum_col:
                        return [[t, sum_col[pre_sum_col] + 1], [b, c]]
                    
                    sum_col[pre_sum_col] = c
                        
        return None

